pkg/util/ip_reservation.go

```diff
@@ -0,0 +1,197 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"math"
+	"net"
+	"sync"
+)
+
+const (
+	ipRangeSize = 29
+	byteMax     = 255
+	ipV4Bits    = 32
+)
+
+var (
+	// step size for IP range increment
+	incrementStep29IPRange = (byte)(math.Exp2(ipV4Bits - ipRangeSize))
+	// mask for IP range
+	ipRangeMask = net.CIDRMask(ipRangeSize, ipV4Bits)
+)
+
+// IPAllocator struct consists of shared resources that are used to keep track of the /29 IPRanges currently reserved by service instances
+type IPAllocator struct {
+	// pendingIPRanges set maintains the set of  IP ranges that have been reserved by the service instances but pending reservation in the cloud instances
+	// The key is a IP range currently reserved by a service instance e.g(192.168.92.0/29). Value is a bool to implement map as a set
+	pendingIPRanges map[string]bool
+
+	// pendingIPRangesMutex is used to synchronize access to the pendingIPRanges set to prevent data races
+	pendingIPRangesMutex sync.Mutex
+}
+
+// NewIPAllocator is the constructor to initialize the IPAllocator object
+// Argument pendingIPRanges map[string]bool is a set of  IP ranges currently reserved by service instances but pending reservation in the cloud instances
+func NewIPAllocator(pendingIPRanges map[string]bool) *IPAllocator {
+	// Make a copy of the pending IP ranges and set it in the IPAllocator so that the caller cannot mutate this map outside the library
+	pendingIPRangesCopy := make(map[string]bool)
+	for pendingIPRange := range pendingIPRanges {
+		pendingIPRangesCopy[pendingIPRange] = true
+	}
+	return &IPAllocator{
+		pendingIPRanges: pendingIPRangesCopy,
+	}
+}
+
+// holdIPRange adds a particular IP range in the pendingIPRanges set
+// Argument ipRange string is an IPV4 range which needs put in pendingIPRanges
+func (ipAllocator *IPAllocator) holdIPRange(ipRange string) {
+	ipAllocator.pendingIPRanges[ipRange] = true
+}
+
+// ReleaseIPRange releases the pending IPRange
+// Argument ipRange string is an IPV4 range which needs to be released
+func (ipAllocator *IPAllocator) ReleaseIPRange(ipRange string) {
+	ipAllocator.pendingIPRangesMutex.Lock()
+	defer ipAllocator.pendingIPRangesMutex.Unlock()
+	delete(ipAllocator.pendingIPRanges, ipRange)
+}
+
+// ValidateCIDR function validates whether a particular cidr is a valid IP range
+// Argument cidr string is is a CIDR range that needs to be validated
+func (ipAllocator *IPAllocator) ValidateCIDR(cidr string) error {
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return err
+	}
+	// The cidr network address needs to be maximum 29 bits
+	cidrSize, _ := ipnet.Mask.Size()
+	if cidrSize > ipRangeSize {
+		return fmt.Errorf("the maximum size of network address in the cidr can be /%d", ipRangeSize)
```

Comment:
"maximum" is sort of misleading because the bigger the number, the smaller the network is.  Maybe instead the message should say "the reserved-ipv4-cidr network size must be at least /29"

cc @bowei for terminology suggestions